pkg/asset/cluster/cluster.go

```diff
@@ -117,17 +117,19 @@ func (c *Cluster) Generate(parents asset.Parents) (err error) {
 		err = errors.Wrap(err, "failed to create cluster")
 	}
 
-	data, err2 := ioutil.ReadFile(stateFile)
-	if err2 == nil {
-		c.FileList = append(c.FileList, &asset.File{
-			Filename: terraform.StateFileName,
-			Data:     data,
-		})
-	} else {
-		if err == nil {
-			err = err2
+	if stateFile != "" {
```

Comment:
This code could use some comments. Reading the code it is not clear to me why the function is not returning when `terraform.Apply` returns an error. Is there a case where `terraform.Apply` returns an empty state file and a nil error? I am afraid that there may be some special considerations being taken in the code that may be lost in future modifications unless there are comments describing why those considerations need to be retained.

Reply:
+1

Reply:
> Reading the code it is not clear to me why the function is not returning when terraform.Apply returns an error. Is there a case where terraform.Apply returns an empty state file and a nil error? 

I've rebased onto master (no conflicts) and updated the commit message to cover both of these questions with e2f919a -> bffc929.

Reply:
I am not seeing where from https://github.com/openshift/installer/commit/3f4fe5741409fde9c183afa180b30fa7c87a0b7c that `terraform.Apply` would return a non-empty state file path and a `nil` error.

Reply:
The only two returns that I see are
1) an empty string and whatever non-nil error was returned by `unpackAndInit`, and
2) the path to the state file and whatever error was returned by `texec.Apply`.

Reply:
I understand better what case we are trying to cover with these changes. I would still love to see an in-code comment explaining that we are trying to preserve the terraform state file even when there was an error applying the terraform.

Reply:
> I would still love to see an in-code comment explaining that we are trying to preserve the terraform state file even when there was an error applying the terraform.

I can add that if pressed.  But if I copy/pasted the ~20 lines of commit message about this in as comments, I think that would make the function less readable.  Is there a problem with digging the motivation out of the commit history via `git blame ...`?

Reply:
I definitely don't want the 20 lines from the commit message copied here as a comment. I was looking for something more along the lines of the following 2 comments.

`Store the error from the apply, but continue with the generation so that the Terraform state file is saved on disk`
https://github.com/openshift/installer/blob/bffc929553ed760f3332f8f734863ae1b840678a/pkg/asset/cluster/cluster.go#L84


`Do not attempt to read the state file if there as an apply error that prevented the state file from being creating`
https://github.com/openshift/installer/blob/bffc929553ed760f3332f8f734863ae1b840678a/pkg/asset/cluster/cluster.go#L87

---
My personal preference would be to move the check for an empty `stateFile` to inside the `if` block when we've already determined that there was an error. That would make the comments less obtrusive to the overall happy-path scenario and make the second comment less important.

```
stateFile, err := terraform.Apply(tmpDir, installConfig.Config.Platform.Name())
if err != nil {
    // Store the error from the apply, but continue with the generation so that the Terraform state file is saved on disk
    err = errors.Wrap(err, "failed to create cluster")
    if stateFile == "" {
        return err
    }
}

data, err2 := ioutil.ReadFile(stateFile)
```

Reply:
> My personal preference would be to move the check for an empty `stateFile` to inside the `if` block...

Ok.  I've done something like that with bffc929 -> 68e550f.  How does that look to you?